src/tokens/non_fungible.rs

```diff
@@ -0,0 +1,600 @@
+        }
+
+        let Some(token) = self.tokens.get_mut(&id) else {
+            return Err(NFTError::TokenNotExists);
+        };
+        let msg_source = msg::source();
+
+        if !(token.owner == msg_source
+            || token.approvals.remove(&msg_source)
+            || Self::inner_allowance(&self.owners, None, token.owner, msg_source))
+        {
+            return Err(NFTError::NotApproved);
+        }
+
+        let from = token.owner;
+
+        token.owner = to;
+
+        self.decrement_balance(from);
+        self.increment_balance(to);
+
+        Ok(NFTTransfer { from, to, id })
+    }
+
+    /// Allows or disallows `operator` to transfer the token with given `id` or
+    /// all [`msg::source()`]'s tokens.
+    ///
+    /// - If `id` is [`Some`], sets an approval only for the token with this
+    /// `id`.
+    ///
+    /// If [`msg::source()`] is an operator of all tokens of the owner of the
+    /// token with given `id`, [`msg::source()`] can set approval any for them.
+    ///
+    /// - If `id` is [`None`], sets an approval for all [`msg::source()`]'s
+    /// tokens.
+    ///
+    /// # Errors
+    /// - [`NFTError::ZeroRecipientAddress`] if `operator` is
+    /// [`ActorId::zero()`].
+    /// - [`NFTError::TokenNotExists`] if the token doesn't exist.
+    /// - [`NFTError::NotApproved`] if [`msg::source()`] isn't the owner of the
+    /// token and operator of token's owner's tokens.
+    pub fn approve(
+        &mut self,
+        operator: Operator,
+        id: Option<Id>,
+        approve: bool,
+    ) -> Result<NFTApproval, NFTError> {
+        if operator.is_zero() {
+            return Err(NFTError::ZeroRecipientAddress);
+        }
+
+        let msg_source = msg::source();
+        let mut owner = msg_source;
+
+        let operators = if let Some(ref unwrapped_id) = id {
+            let current_owner = self.internal_owner_of(unwrapped_id)?;
+
+            if current_owner != owner {
+                if !self.allowance(current_owner, owner, None) {
+                    return Err(NFTError::NotApproved);
+                }
+
+                owner = current_owner;
+            }
+
+            &mut self.tokens.get_mut(unwrapped_id).unwrap().approvals
+        } else {
+            &mut self.owners.entry(owner).or_default().operators
+        };
+
+        if approve {
+            operators.insert(operator);
+        } else {
+            operators.remove(&operator);
+        }
+
+        Ok(NFTApproval {
+            owner,
+            operator,
+            id,
+            approved: approve,
+        })
+    }
+
+    fn balance_of_mut(&mut self, owner: Owner) -> &mut Amount {
+        &mut self.owners.entry(owner).or_default().balance
+    }
+
+    fn increment_balance(&mut self, owner: Owner) {
+        *self.balance_of_mut(owner) += Amount::one();
+    }
+
+    fn decrement_balance(&mut self, owner: Owner) {
+        *self.balance_of_mut(owner) -= Amount::one();
```

Comment:
Is it safe?

Reply:
Yes. All methods check whether a token has the owner before calling this method. This prevents an overflow because if the owner has the token, its balance must be greater than 0.